src/main/java/de/tum/in/www1/artemis/web/rest/TextSubmissionResource.java

```diff
@@ -171,68 +172,51 @@ private void hideDetails(@RequestBody TextSubmission textSubmission) {
+        if (assessedByTutor) {
+            User user = userService.getUserWithGroupsAndAuthorities();
 
-        return ResponseEntity.ok().body(
-
-            // Participations for Exercise
-            participationService.findByExerciseIdWithEagerSubmissions(exerciseId)
-                .stream()
-                .peek(participation -> participation.getExercise().setParticipations(null))
-
-                // Map to Latest Submission
-                .map(Participation::findLatestTextSubmission)
-                .filter(Objects::nonNull)
+            return ResponseEntity.ok().body(
+                textSubmissionService.getAllTextSubmissionsByTutorForExercise(exerciseId, user.getId())
+            );
+        }
 
-                // if submittedOnly, need to check if submitted, else just continue (!submittedOnly == true)
-                .filter(submission -> !submittedOnly || submission.isSubmitted())
+        List<Participation> participations = participationService.findByExerciseIdWithEagerSubmissions(exerciseId);
+        List<TextSubmission> textSubmissions = new ArrayList<>();
 
-                // Load Result for Submission
-                .peek(textSubmission -> {
-                    Hibernate.initialize(textSubmission.getResult()); // eagerly load the association
-                    if (textSubmission.getResult() != null) {
-                        Hibernate.initialize(textSubmission.getResult().getAssessor());
-                        textSubmission.getResult().getAssessor().setGroups(null);
-                    }
-                })
+        for (Participation participation : participations) {
+            TextSubmission textSubmission = participation.findLatestTextSubmission();
 
-                // Convert Stream to List to Match Return Type
-                .collect(Collectors.toList())
-        );
-    }
+            // if submittedOnly, need to check if submitted, else just continue (!submittedOnly == true)
+            if (textSubmission == null || (submittedOnly && !textSubmission.isSubmitted())) {
+                continue;
+            }
 
-    /**
-     * GET  /text-submissions-assessed-by-tutor : get all the textSubmissions assessed by the current tutor.
-     *
-     * @return the ResponseEntity with status 200 (OK) and the list of textSubmissions in body
-     */
-    @GetMapping(value = "/exercises/{exerciseId}/text-submissions-assessed-by-tutor")
-    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
-    @Transactional(readOnly = true)
-    public ResponseEntity<List<TextSubmission>> getAllTextSubmissionsByTutor(@PathVariable Long exerciseId) {
-        log.debug("REST request to get all TextSubmissions filtered by assessed by tutor");
-        Exercise exercise = exerciseService.findOneLoadParticipations(exerciseId);
-        User user = userService.getUserWithGroupsAndAuthorities();
+            Hibernate.initialize(textSubmission.getResult()); // eagerly load the association
```

Comment:
these calls to Hibernate.initialize might be expensive if there are many participations (assume in EIST we have 2000 of them). Can we refactor this and instead eagerly load them together with the participations? I.e. change the logic of line 199 and use different database calls depending on the two parameters? Also there is too much logic involve here for a simple resource, so I suggest to move this into the corresponding service.

Reply:
I moved to a service and improved the query